### ISSUE TITLE
DRM and clear lead causes glitch

### ISSUE THREAD

**Issue by kgrevehagen** [google/ExoPlayer#6751]
### [REQUIRED] Searched documentation and issues
ExoPlayer Javadoc and code, exoplayer.dev, issue tracker, stack overflow.

### [REQUIRED] Question
When using DRM(we have tested audio only with CMAF/CENCCENC/Widevine) and 20 seconds clear lead, we are noticing an audible glitch on some older devices(like the LG G3, Samsung S7, Huawei P10 Lite etc) when the native `MediaDrm::provideKeyResponse`(https://android.googlesource.com/platform/frameworks/base/+/b267554/media/java/android/media/MediaDrm.java#383) is called.

We believe this is because both the `AudioTrack::write` and the `MediaDrm::provideKeyResponse` is blocking on the same thread, thus causing the glitch and a callback to `onAudioUnderrun`. We also tried calling `MediaDrm::provideKeyResponse` on a separate thread, without success.

We found that increasing `DefaultAudioSink::MIN_BUFFER_DURATION_US` to e.g. 2500000 us (2500 ms) fixed the problem.

What implications would increasing `DefaultAudioSink::MIN_BUFFER_DURATION_US` from 250 ms to 2500 ms(or even higher) have? We would need to change the `DefaultAudioSink::MAX_BUFFER_DURATION_US` similarly, I suppose?

`MediaDrm::provideKeyResponse` timing on LG G3: approx 2500 ms
`MediaDrm::provideKeyResponse` timing on Pixel 3: < 100 ms
 
Do you know of any other solution to this issue?

### A full bug report captured from the device
N/A

### Link to test content
N/A

**Comment by andrewlewis** [comment]
Is this related to #6459 or a separate issue? Thanks.

**Comment by kgrevehagen** [comment]
> Is this related to #6459 or a separate issue? Thanks.

This is a separate issue. It is very closely related to #6344. I think the underlying issue is the same, just the use case is a bit different.
This is 100% reproducible on the affected devices.

I think what @ojw28 mentions here (https://github.com/google/ExoPlayer/issues/6344#issuecomment-531517852) is very interesting, regarding single threaded trusted execution environment.

There has to be a way around this, or at least to identify the cases so we can turn off clear lead.

Let me know if there is anything else you need answers about regarding this issue.
Thanks for looking into this!

**Comment by kgrevehagen** [comment]
@andrewlewis Any news or comments on this issue? Basically, clear lead is not working without glitch on many devices, so I think this is a pretty serious issue, do you agree?

**Comment by andrewlewis** [comment]
@kgrevehagen Agreed this looks very similar to #6344.

Regarding AudioTrack buffers: the buffer is drawn from shared memory so my main concern would be that you may see issues with the AudioTrack failing to initialize if you make the buffer significantly larger. Larger buffer sizes also increase latency but this may not matter for your use case (if the content is live the AudioTrack buffer will 'delay' playout; and if you changing playback speed the latency of those changes will increase when you increase audio audio track buffer size).

Please could you provide the additional information requested in https://github.com/google/ExoPlayer/issues/6344#issuecomment-532084342? I've assigned this to @rrfrias for further investigation as that issue has been closed.

**Comment by rrfrias** [comment]
@kgrevehagen can you let us know what android builds the devices you mentioned are running? We have made some improvements in concurrency in android 10. It might also be more pronounced on devices with slower processors. 

I assume you are not seeing the problem on a pixel 3. Is that running 10 or earlier?

**Comment by rrfrias** [comment]
If the problematic devices are running android O mr1 (8.1) or later, can you please include a bugreport as well.

**Comment by kgrevehagen** [comment]
Information requested in comment in #6344:

> (a) let us know which devices (make/model) and OS builds/releases you experience the problems with. If there are those that do not exhibit the problem, do let us know that as well.
- I'm reproducing it 100% on Huawei P10 Lite on Android 8.0.0 and LG G3 S (LG-D722) on Android 5.0.2, and our metrics shows that it happens on a lot of other devices as well(mostly low end), though I have not personally tested this. It is not happening on Pixel 3 on Android 10.

> (b) can you share a bugreport taken on one of the affected devices, after this problem occurs.
- I think it is much better if you are able to test this yourself and see it in action, but I can provide one if still needed.

> (c) if possible share a test app, and a stream for us to test.
- If I remember correctly, this is reproducible on the mentioned devices on the DRM content already available in the sample app if you activate clear lead. If this is not working, I will provide you with test content. I will get back to you on this once I have tested it thoroughly again.

Let me know if you need any more information.

**Comment by jt64** [comment]
Given that these older OS versions are probably not receiving any more updates, fixes to the framework realistically aren't going to happen. The options you have would be to increase the audio buffer size as you have found, or to request the license before starting playback.
